src/loki/search.cc

```diff
@@ -120,21 +120,42 @@ struct candidate_t {
     return sq_distance < c.sq_distance;
   }
 
-  PathLocation::SideOfStreet
-  get_side(const PointLL& original, double sq_distance, double sq_tolerance) const {
-    // its so close to the edge that its basically on the edge
-    if (sq_distance < sq_tolerance) {
+  PathLocation::SideOfStreet get_side(const PointLL& original,
+                                      double sq_distance,
+                                      double sq_tolerance,
+                                      double sq_max_distance) const {
+    // point is so close to the edge that its basically on the edge,
+    // or its too far from the edge that we shouldn't use it to determine side of street
+    if (sq_distance < sq_tolerance || sq_distance > sq_max_distance) {
       return PathLocation::SideOfStreet::NONE;
     }
 
-    // get the side TODO: this can technically fail for longer segments..
-    // to fix it we simply compute the plane formed by the triangle
-    // through the center of the earth and the two shape points and test
-    // whether the original point is above or below the plane (depending on winding)
+    // get the angle of the tangent line to the shape at this point
     auto& shape = edge_info->shape();
-    LineSegment2<PointLL> segment(shape[index], shape[index + 1]);
-    return (segment.IsLeft(original) > 0) == edge->forward() ? PathLocation::SideOfStreet::LEFT
-                                                             : PathLocation::SideOfStreet::RIGHT;
+    auto tang_angle =
```

Comment:
we should take this opportunity to clean up a little bit here. there are performance implications. see every time we do this we have to pull out the shape and do spherical math on it. its not cheap. you'll notice that above in the heading function we are doing the same thing (though only when the person asks for a heading filter).

what i think we should do is first grab the tangent angle and store it as part of the candidate. ie do it only once. and then depending on `edge->forward()` we use it or it `+180` to do both the side of street and the heading filter.

Reply:
👍  done